python/salpytools/states.py

```diff
@@ -34,33 +34,33 @@
 state_enumeration["STANDBY"] = 1
 state_enumeration["DISABLED"] = 2
 state_enumeration["ENABLED"] = 3
-state_enumeration["FAULT"]   = 4
+state_enumeration["FAULT"] = 4
 state_enumeration["INITIAL"] = 5
-state_enumeration["FINAL"] =   6
+state_enumeration["FINAL"] = 6
 
 # This matrix expresses valid transitions and is reproduced in code afterwards.
 #
-#    \NEXT STATE
-#STATE\
-#      \ |Offline |Standby |Disabled|Enabled |Fault   |Initial |Final   |
-#------------------------------------------------------------------------ 
-#Offline | TRUE   | TRUE   |        |        |        |        |  TRUE  |
-#------------------------------------------------------------------------
-#Standby |  TRUE  | TRUE   |  TRUE  |        |  TRUE  |        |  TRUE  |
-#------------------------------------------------------------------------
-#Disable |        |  TRUE  |  TRUE  |  TRUE  |  TRUE  |        |        |
-#------------------------------------------------------------------------
-#Enable  |        |        |  TRUE  |  TRUE  |  TRUE  |        |        |
-#------------------------------------------------------------------------
-#Fault   |        |        |        |        |  TRUE  |        |        |
-#------------------------------------------------------------------------
-#Initial |        |  TRUE  |        |        |        | TRUE   |        |
-#------------------------------------------------------------------------
-#Final   |        |        |        |        |        |        | TRUE   |
-#------------------------------------------------------------------------
+#     \NEXT STATE
+# STATE\
+#       \ |Offline |Standby |Disabled|Enabled |Fault   |Initial |Final   |
+# ------------------------------------------------------------------------
+# Offline | TRUE   | TRUE   |        |        |        |        |  TRUE  |
+# ------------------------------------------------------------------------
+# Standby |  TRUE  | TRUE   |  TRUE  |        |  TRUE  |        |  TRUE  |
+# ------------------------------------------------------------------------
+# Disable |        |  TRUE  |  TRUE  |  TRUE  |  TRUE  |        |        |
+# ------------------------------------------------------------------------
+# Enable  |        |        |  TRUE  |  TRUE  |  TRUE  |        |        |
+# ------------------------------------------------------------------------
+# Fault   |        |        |        |        |  TRUE  |        |        |
+# ------------------------------------------------------------------------
+# Initial |        |  TRUE  |        |        |        | TRUE   |        |
+# ------------------------------------------------------------------------
+# Final   |        |        |        |        |        |        | TRUE   |
+# ------------------------------------------------------------------------
```

Comment:
It might be nice to express this as a string (or matrix), with the commands for the allowed transitions in the boxes instead of TRUE.  Code could then parse the string/matrix, generating `next_state`, `state_enumeration`, and `state_matrix` from the same authoritative source.  As currently written, changes to the state machine may require multiple synchronized modifications that could introduce errors.

Furthermore, because `next_state` has no knowledge of the preceding state, I believe that the code allows, for example, "ENTER_CONTROL" to be issued in DISABLED state to go to STANDBY state, while that command is only supposed to be valid in OFFLINE state.

Reply:
This code is the same used by DMCS and I am reluctant to change it, without Jim agreeing to change it the same way.

Reply:
https://github.com/lsst/ctrl_iip/blob/master/python/lsst/iip/toolsmod.py#L70